src/nvim/buffer.c

```diff
@@ -3069,9 +3069,13 @@ static bool ti_change(char_u *str, char_u **last)
 /// Set current window title
 void resettitle(void)
 {
-  ui_call_set_title(cstr_as_string((char *)lasttitle));
-  ui_call_set_icon(cstr_as_string((char *)lasticon));
-  ui_flush();
+  if (p_icon) {
```

Comment:
These conditions should be in the TUI not here, it breaks other UIs.

Reply:
I thought these options applied to all UIs. It just raises UI events, how does it break them?

Reply:
For a terminal it makes sense to either set the title or retain the old title (set by the emulator or the shell). For the UI it makes no sense as there is no "old title" to override/reset. vim handles this relatively reasonably by setting `'title'` by default in gvim, but not (necessarily) in terminal vim. But as nvim can't let the default be affected by the ui/terminal type one way is to let UIs always get the title, but let the TUI be controlled by default-off option.

Alternatively, we could `set title` by default also in TUI, but I have no idea how many issues opened by upset users that would lead to.

Reply:
Ah, got it. We didn't have these conditions before, so we could just default title=on , right?

Reply:
After checking, we already have them in `maketitle()`, so these conditions are  redundant. (and master before this doesn't set terminal title with `-u NONE`) . So the question still stands. _Personally_ I'm perfectly fine with `set title` by default, but if `guicursor` is any precedent, prepare for responses like "nvim set the title and didn't restore it, it broke my terminal setup".  Your call, I guess you will be the one responding to them. :) But there is a third way, we can hack the vim behavior a simpler way; if nvim is started with `--headless` or `--embed`, then `set title` by default.

Reply:
> we already have them in maketitle(), so these conditions are redundant

But `resettitle()` is not only called from `maketitle()`. 

Vim doesn't have these conditions in `resettitle`, it only calls `mch_settitle`, which is driven by the state of `lasttitle`/`lasticon`, which are set by `maketitle()`. `mch_settitle()` doesn't exist in Nvim, it was replaced by `ui_call_set_*`, i.e., it's delegated to the UI.

I actually don't know what the implications of the conditions are, and if it works ok without them, might as well remove them as they were before. The logic is just crazy.